### PR TITLE
✨ encodeURIComponent

### PR DESCRIPTION
**Pull request by Vectorized**
## Description

Continued from #1097.

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `forge fmt`?
- [x] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

**Comment by johnnyshankman** [comment]
Happy to have contributed! Thanks for the help @Vectorized

File: test/LibString.t.sol [hunk 1]
```diff
@@ -1097,6 +1097,182 @@ contract LibStringTest is SoladyTest {
...
+        assertEq(encoded, _encodeURIComponentOriginal(s));
+        if (_randomChance(8)) {
+            assertEq(_decodeURIComponentOriginal(encoded), s);
+        }
+        _checkMemory(encoded);
+        assertEq(keccak256(bytes(s)), hashBefore);
+    }
+
+    // Original implementation of `_encodeURIComponentOriginal`
+    // credit to John Shankman aka White Lights - johnny@white-lights.net (whitelights.eth).
```
Comment (johnnyshankman) [inline test/LibString.t.sol:1139]: <3

File: src/utils/LibString.sol [hunk 2]
```diff
@@ -1098,6 +1098,39 @@ library LibString {
...
+                    continue
+                }
+                mstore8(o, c)
+                o := add(o, 1)
+            }
+            mstore(result, sub(o, add(result, 0x20))) // Store the length.
+            mstore(o, 0) // Zeroize the slot after the string.
+            mstore(0x40, add(o, 0x20)) // Allocate memory.
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Vectorized) [reply src/utils/LibString.sol:1132]: Opcodes are like the movements of martial arts and fingerstyle guitar.
Study, practice, experiment.
Then combine them to create art.

Marcin is a great inspiration.

https://youtu.be/0FImBCNRGxo?si=hMxZMapKPksYmrFv

[the remark this answers is not in the document]
Comment (johnnyshankman) [reply src/utils/LibString.sol:1132]: big thank you. this truly seems like a whole different mindset. once you have a working implementation, it's then about breaking it down into pieces of byte-wise memory and trying to optimize on the mload/mstore level.

this is fun cause now i have a little test. if i can manage to optimize my original function the same way you have (or similarly) i'll know i'm starting to get the hang of it.

thank you vec!

[the remark this answers is not in the document]
Comment (johnnyshankman) [reply src/utils/LibString.sol:1132]: sidenote: i can play fingerstyle guitar and this somehow seems harder lmao. marcin is a g. you ever listen to andy mckee or chet atkins? could listen for hours.

[the remark this answers is not in the document]
Comment (atarpara) [reply src/utils/LibString.sol:1132]: @johnnyshankman, may I know why you included `_` in this encoding method? According to the [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#description), it should be excluded.

[the remark this answers is not in the document]
Comment (Vectorized) [reply src/utils/LibString.sol:1132]: Good spot.

[the remark this answers is not in the document]
Comment (johnnyshankman) [reply src/utils/LibString.sol:1132]: good catch indeed, oversight on my part.

Screenshot 2024-10-03 at 4 16 46 PM

**Pull request merged**